dwave/gate/operations/base.py

```diff
@@ -41,22 +41,34 @@
     overload,
 )
 
+import numpy as np
+
 from dwave.gate.circuit import Circuit, CircuitContext, ParametricCircuit
 from dwave.gate.mixedproperty import mixedproperty
-from dwave.gate.primitives import Qubit
-from dwave.gate.registers.registers import Variable
+from dwave.gate.primitives import Bit, Qubit
+from dwave.gate.registers.registers import ClassicalRegister, Variable
+from dwave.gate.tools.samples import sample
 from dwave.gate.tools.unitary import build_controlled_unitary, build_unitary
 
 if TYPE_CHECKING:
     from numpy.typing import NDArray
 
+# define an integer value for an arbitrary number of qubits;
+# used for clarity in variable qubit number operations
+ANY = -1
```

Comment:
Also a good idea, but I'm not sure if it's necessary here. I'm quite OK with having -1 represent an arbitrary number of qubits, but I'll play around a bit with it.

> Or better yet to handle arithmetic comparison as well, perhaps raising (meaningful) error when ANY compared to number of qubits, you can implement a new type.

I think this could be the better option.

Reply:
@randomir Added an ``AnyClass`` to represent an arbitrary number of qubits instead ([see this commit](https://github.com/dwavesystems/dwave-gate/pull/12/commits/17802ea1b90c439de8813a73ebf4eaf1602b323e#diff-295c264653199148252d266a754a5553c61b034111d8504405adc2ad754ab60eR69-R105)). Let me know what you think. It inherits from int (to keep `_num_qubits` as ints`, while disabling arithmetics and comparisons.

Reply:
Seems like a total overkill, IMO. :smile: Also, I don't like `AnyClass` as `int` subclass (breaks Liskov).

I might not have the bigger picture here, but from the code, it looks like `ANY` will be used internally only (it won't seep outside object methods), so `None` would work just as well (special meaning, arithmetic/comparison disabled). OTOH, the same stands for the original `-1`, except that concrete "forbidden" int value with special meaning is too easy to mutate by accident (hence loosing the "special" meaning). If you need to keep `None` reserved for a different meaning (like value not assigned), then a simple `ANY = object()` is the way to go (creates "new `None`").

Reply:
Fair enough. I might've gone a bit unnecessarily far on this one. 😅 I'm not too worried about Liskov in this case though, since it's more of one-off/single-instance used class, but I definitely get the point. The nice thing with using this class (or -1 for that matter) is that it stays an int, which is handy both for type-hinting and instance-checking. 

Personally, I have a slight preference to using either of those two alternatives, but I don't really have any strong feelings against just using a separate object placeholder (like `object()` or `...`) instead.

> it looks like ANY will be used internally only (it won't seep outside object methods)

That's correct. `None`, or even `...`, would definitely work, but maybe just not _as_ well.

Reply:
I see type-hinting more as convenience than something driving the design. But if that's important here, you can create a new type (`Optional`-like), e.g. `Union[int, Any, None]`.

Reply:
Wouldn't it be more explicit to do
```python
class MyClass:
    def is_variable_length(self) -> bool:
        return True
    def num_qubits(self) -> int:
        raise TypeError("I don't have a fixed length")
```
that way you can get around a lot of these typing issues. Though of course it possibly ends up being too verbose.

Reply:
I went with @arcondello's suggestion (more-or-less), and I reworked the way `_num_qubits`, `_num_control`, `_num_targets`, and `_num_params` are declared. This way, `None` represents "any" number of qubits and will raise the appropriate error if accessed before having been set when initialized.